dhcpv4/nclient4/lease.go

```diff
@@ -0,0 +1,227 @@
+	rstr += fmt.Sprintf("Id options:\n%v", lease.IDOptions)
+	rstr += fmt.Sprintf("ACK options:\n%v", lease.AckOptions)
+	return rstr
+}
+
+// WithClientIDOptions configures a list of DHCPv4 option code that DHCP server
+// uses to identify client, beside the MAC address.
+func WithClientIDOptions(cidl dhcpv4.OptionCodeList) ClientOpt {
+	return func(c *Client) (err error) {
+		c.clientIDOptions = cidl
+		return
+	}
+}
+
+// WithApplyLeaseHandler specifies a handler function which is called when
+// Client.ApplyLease() is called; without this, a default handler function is called.
+// the default handler will add/remove the assigned address to/from the binding interface;
+// bool parameter is true when lease is applied, false when lease is released
+func WithApplyLeaseHandler(h func(DHCPv4ClientLease, bool) error) ClientOpt {
+	return func(c *Client) (err error) {
+		c.leaseApplyHandler = h
+		return
+	}
+}
+
+//default lease apply handler
+//add/remove address to/from binding interface
+func defaultLeaseApplyHandler(l DHCPv4ClientLease, enable bool) error {
+	link, err := netlink.LinkByName(l.IfName)
+	if err != nil {
+		return err
+	}
+	plen, _ := l.AssignedIPMask.Size()
+	prefixstr := fmt.Sprintf("%v/%v", l.AssignedIP, plen)
+	naddr, err := netlink.ParseAddr(prefixstr)
+	if err != nil {
+		return err
+	}
+	if enable {
+		err = netlink.AddrReplace(link, naddr)
+
+	} else {
+		err = netlink.AddrDel(link, naddr)
+	}
+	return err
+
+}
+
+//ApplyLease apply/unapply the lease, call the c.leaseApplyHandler
+func (c *Client) ApplyLease(enable bool) error {
+	if c.lease == nil {
+		return fmt.Errorf("no lease to apply")
+	}
+	return c.leaseApplyHandler(c.GetLease(), enable)
+}
+
+//GetLease return the lease
+func (c *Client) GetLease() (clease DHCPv4ClientLease) {
+	clease = *c.lease
+	clease.MACAddr = c.ifaceHWAddr
+	clease.IfName = c.ifName
+	clease.ServerAddr = *c.serverAddr
+	return
+}
+
+// RequestSavingLease completes DORA handshake and store&apply the lease
+//
+// Note that modifiers will be applied *both* to Discover and Request packets.
+func (c *Client) RequestSavingLease(ctx context.Context, modifiers ...dhcpv4.Modifier) (offer, ack *dhcpv4.DHCPv4, err error) {
+	offer, err = c.DiscoverOffer(ctx, modifiers...)
+	if err != nil {
+		err = fmt.Errorf("unable to receive an offer: %w", err)
+		return
+	}
+
+	// TODO(chrisko): should this be unicast to the server?
+	request, err := dhcpv4.NewRequestFromOffer(offer, dhcpv4.PrependModifiers(modifiers,
+		dhcpv4.WithOption(dhcpv4.OptMaxMessageSize(MaxMessageSize)))...)
+	if err != nil {
+		err = fmt.Errorf("unable to create a request: %w", err)
+		return
+	}
+
+	ack, err = c.SendAndRead(ctx, c.serverAddr, request, nil)
+	if err != nil {
+		err = fmt.Errorf("got an error while processing the request: %w", err)
+		return
+	}
+	//save lease
+	c.lease = &DHCPv4ClientLease{}
```

Comment:
I'm not sure I understand entirely why DHCPv4ClientLease exists. I would think that the ACK packet *is* the lease, and we end up with 

```go
type DHCPv4ClientLease struct {
  ACK *DHCPv4
  CreationTime time.Time
}
```

because I don't think the information you pulled out is all that users are going to want to use from the ACK packet.

Reply:
yes, most of part of lease come from ACK, beside the the ID options used during DORA, so maybe I could change it to sth like following:

```
type DHCPv4ClientLease struct {
  ACK *DHCPv4
  CreationTime time.Time
  IDOptions      dhcpv4.Options
}
```

Reply:
I'd appreciate that, yeah! I'm not sure I fully get the IDOptions thing yet. What ID options are these, specifically, that aren't in the ack and aren't known to this library already?

Reply:
rfc2131, section 4.2, specifies that client-id option61 could be used to identify client, quote below
`
The client MAY choose to explicitly provide
   the identifier through the 'client identifier' option.  If the client
   supplies a 'client identifier', the client MUST use the same 'client
   identifier' in all subsequent messages, and the server MUST use that
   identifier to identify the client.`
then in section 4.3.1's table, client-id is marked as "MUST NOT" for DHCPACK, meaning ACK will not include client-id; so in such case, client need to remember the client-id used during DORA, and use the same value in release, otherwise server won't be able to identify the lease that release targets.  

So the idea of IDOptions is meant to be a generic way for option like client-id, for example a DHCP server might require using class-id(option60) + client-id to identify client, then IDOptions will include option 60 and 61;

Reply:
Makes sense.